DDC-2332: Ensure managed entities are always tracked by UOW

bigfoot90 opened the pull request:
This is a rebase of MR https://github.com/doctrine/orm/pull/7407 on `2.9.x` branch.
Fixes https://github.com/doctrine/orm/issues/3037

olsavmic committed:
Fix mapped superclass missing in discriminator map

bigfoot90 committed:
DDC-2332: Ensure managed entities are always tracked by UOW

phansys commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -318,6 +318,15 @@ class UnitOfWork implements PropertyChangedListener
+     *
+     * DDC-2332 / #3037
+     *
+     * @var array
```
```suggestion
     * @psalm-var array<string, object>
```

phansys commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -318,6 +318,15 @@ class UnitOfWork implements PropertyChangedListener
+     *
+     * @var array
+     */
+    private $oidMap = array();
```
```suggestion
    private $oidMap = [];
```

phansys commented on tests/Doctrine/Tests/ORM/Functional/IdentityMapTest.php:
```diff
@@ -257,4 +258,49 @@ public function testCollectionValuedAssociationIdentityMapBehaviorWithRefresh():
+        $this->_em->flush();
+
+        $articles = [];
+        for ($i = 0; $i < 100; $i++) {
```
```suggestion
        for ($i = 0; $i < 100; ++$i) {
```

phansys commented on tests/Doctrine/Tests/ORM/Functional/IdentityMapTest.php:
```diff
@@ -257,4 +258,49 @@ public function testCollectionValuedAssociationIdentityMapBehaviorWithRefresh():
+            $this->_em->persist($article);
+            $this->_em->flush();
+            $this->_em->clear();
+            $articles [] = $article;
```
```suggestion
            $articles[] = $article;
```

phansys commented on tests/Doctrine/Tests/ORM/Functional/IdentityMapTest.php:
```diff
@@ -257,4 +258,49 @@ public function testCollectionValuedAssociationIdentityMapBehaviorWithRefresh():
+        gc_collect_cycles();
+        $keep = [];
+
+        for ($x = 0; $x < 1000; $x++) {
```
```suggestion
        for ($x = 0; $x < 1000; ++$x) {
```

greg0ire commented:
`vendor/bin/phpcbf` can help with the CS issues

bigfoot90 commented:
Thanks, fixed all cs

bigfoot90 commented:
@greg0ire Friendly ping

qkdreyer commented:
Any updates on this ?

bigfoot90 commented:
Rebased on `2.10.x`

beberlei commented:
I am skeptical. Can you explain the test a little better via comments? Its doing s lot, not sure where exactly and what fails. Is the merge necessary to cause this?

bigfoot90 commented:
Rebased on `2.10.x` again and added some comments on test as suggested.

This same test `testObjectIdCollision` copy-pasted on  branch `2.10.x` is failing on iteration 60, but your could be different.

bigfoot90 committed:
Last comment

carlos-granados commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -1541,6 +1551,8 @@ public function addToIdentityMap($entity)
         $idHash    = implode(' ', $identifier);
         $className = $classMetadata->rootEntityName;
 
+        $this->oidMap[$oid] = $entity;
```
Instead of saving the whole document, just save a reference to it. This will also stop it from being garbage collected and uses much less memory

$this->oidMap[$oid] = &$entity;

mpdude commented:
@bigfoot90 maybe you want to join #10791
